Raven.Client.Lightweight/Document/InMemoryDocumentSessionOperations.cs

```diff
@@ -693,34 +693,34 @@ public void Store(object entity)
         {
             string id;
             var hasId = GenerateEntityIdOnTheClient.TryGetIdFromInstance(entity, out id);
-            StoreInternal(entity, null, null, forceConcurrencyCheck: hasId == false);
+            StoreInternal(entity, null, null, forceConcurrencyCheck: hasId == false, disableConcurrencyCheck: false);
         }
 
         /// <summary>
         /// Stores the specified entity in the session. The entity will be saved when SaveChanges is called.
         /// </summary>
         public void Store(object entity, Etag etag)
         {
-            StoreInternal(entity, etag, null, forceConcurrencyCheck: true);
+            StoreInternal(entity, etag, null, forceConcurrencyCheck: etag != null, disableConcurrencyCheck: etag == null);
         }
 
         /// <summary>
         /// Stores the specified entity in the session, explicitly specifying its Id. The entity will be saved when SaveChanges is called.
         /// </summary>
         public void Store(object entity, string id)
         {
-            StoreInternal(entity, null, id, forceConcurrencyCheck: false);
+            StoreInternal(entity, null, id, forceConcurrencyCheck: false, disableConcurrencyCheck: false);
         }
 
         /// <summary>
         /// Stores the specified entity in the session, explicitly specifying its Id. The entity will be saved when SaveChanges is called.
         /// </summary>
         public void Store(object entity, Etag etag, string id)
         {
-            StoreInternal(entity, etag, id, forceConcurrencyCheck: true);
+            StoreInternal(entity, etag, id, forceConcurrencyCheck: etag != null, disableConcurrencyCheck: etag == null);
         }
 
-        private void StoreInternal(object entity, Etag etag, string id, bool forceConcurrencyCheck)
+        private void StoreInternal(object entity, Etag etag, string id, bool forceConcurrencyCheck, bool disableConcurrencyCheck)
```

Comment:
Please see the **_whole_** PR. That's what I did in the last commit:

https://github.com/ayende/ravendb/pull/4197/commits/d33dbb08be41e6f55f108fd0e6ee25bd989fde1b

I haven't merged the commits so that it's easier for you to follow what I've changed.